OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,980 @@
+			// If bot have conyard and MCV, bot should wait for MCV deploying before build another MCV.
+			if ((conyardNum <= 0 && mcvNum > 1) || (conyardNum > 0 && mcvNum > 0))
+				return;
+
+			if (conyardNum + mcvNum >= Info.MinimumConstructionYardCount)
+				return;
+
+			// We have MCV in production queue, let's wait.
+			if (mcvFactories.Actors
+				.Any(a => !a.IsDead && a.TraitsImplementing<ProductionQueue>().Any(t => t.Enabled && t.AllQueued().Any(q => Info.McvTypes.Contains(q.Item)))))
+				return;
+
+			// We have MCV in production queue, let's wait.
+			if (player.PlayerActor.TraitsImplementing<ProductionQueue>()
+				.Any(t => t.Enabled && t.AllQueued().Any(q => Info.McvTypes.Contains(q.Item))))
+				return;
+			var unitBuilder = requestUnitProduction.FirstEnabledTraitOrDefault();
+			if (unitBuilder == null)
+				return;
+			var mcvType = Info.McvTypes.Random(world.LocalRandom);
+
+			// Make sure we only request one MCV at a time.
+			if (unitBuilder.RequestedProductionCount(bot, mcvType) <= 0)
+				unitBuilder.RequestUnitProduction(bot, mcvType);
+		}
+
+		void DeployMcvs(IBot bot, bool chooseLocation)
+		{
+			var newMCVs = world.ActorsHavingTrait<Transforms>()
+				.Where(a => a.Owner == player && a.IsIdle && Info.McvTypes.Contains(a.Info.Name));
+
+			foreach (var mcv in newMCVs)
+				DeployMcv(bot, mcv, chooseLocation);
+		}
+
+		void UnDeployConyard(IBot bot)
+		{
+			/*
+			 * Current BaseBuilderBotModule has a very weak control of BO (Build Orders), produce randomly and cannot cooperate well,
+			 * which makes expanding with only 1 MCV performs terrible: bot's necessary base building always get interrupted and spend too many time on road.
+			 *
+			 * Therefore, the considerations of undeploy and move MCV are simple and conservative, at least for now.
+			 *
+			 * TODO: This should be tied deeply with BaseBuilderBotModule in order to keep better control on having only 1 MCV:
+			 * undeploy should affect BaseBuilderBotModule's BO, while BaseBuilderBotModule has some control on MCV's expansion,
+			 * in order to make expansion worth the time and risk.
+			 */
+			if (firstUndeploy)
+			{
+				/*
+				 * On the first undeploy, we only consider if bot player encounter the early game rush, when bot player
+				 * has less than 2 production buildings (barrack, war factory, etc) or no refinery, MCV won't move and leave everything behind.
+				 */
+				if (world.ActorsHavingTrait<Building>().Count(a => a.Owner == player && Info.ProductionTypes.Contains(a.Info.Name)) >= 2
+					&& world.ActorsHavingTrait<Refinery>().Any(a => a.Owner == player && Info.RefineryTypes.Contains(a.Info.Name)))
+				{
+					var idleconyards = constructionYards.Actors.Where(a => !a.IsDead).ToList();
+
+					if (idleconyards.Count > 0)
+					{
+						bot.QueueOrder(new Order("DeployTransform", idleconyards[0], true));
+						allowfallback = false;
+					}
+				}
+
+				firstUndeploy = false;
```

Comment:
```suggestion
					var idleconyards = constructionYards.Actors.FirstOrDefault(a => !a.IsDead);
					if (idleconyards != null)
					{
						bot.QueueOrder(new Order("DeployTransform", idleconyards, true));
						allowfallback = false;
						firstUndeploy = false;
					}
				}

```

Maybe let's set firstUndeploy to false only if we actually have a conyard to undeploy?

Reply:
This still seems flawed, this basically makes sure the AI undeploys and moves mcv as soon as the criteria is reached. It is the playstyle in Tiberium Dawn, but not in RA where you want to build more buildings first. In RA undeplying this early is an a basepushing all-in move, besides in more packed games the AI tends to just lose it's first mcv by driving through an enemies base

Reply:
And in TD all AI's do undeploy their mcv's basically at the same time

Reply:
> this basically makes sure the AI undeploys and moves mcv as soon as the criteria is reached

So I only allow  `firstUndeploy` to be check for only once: if bot meet early game rush and lose buildings, then it loses the chance to first undeploy forever -- but it is better to be safe than sorry, it can also use another MCV to expand.

And also it is not "as soon as", because we have 2000+ ticks in `MoveConyardInterval` so actual time will be quite delayed, unless BaseBuilder take the control of expansion timing (as soon as building is placed).

> In RA undeplying this early is an a basepushing all-in move

The timimg I use is 4200 (2pp+2ref+1bar+1wf), I think it is the best timing I see for RA bot to perform a single MCV expansion with all basic base buildings done. Plus we can add some random to `FirstMoveConyardTicks`, like we can let AI choose among 3500(2pp+1ref+1bar+1wf), 4200(2pp+2ref+1bar+1wf), 5200(2pp+2ref+1bar+1wf+1sd)

> And in TD all AI's do undeploy their mcv's basically at the same time
1. An issue is on performing a single MCV expansion, we have to be percise in time to get the best eco effect, so I set the timing here as time after building 2pp+2ref+1bar+1wf as well. 
2. We can also use random to `FirstMoveConyardTicks` here, but need more tests to find a set of value of timing on moving the only MCV.